Issue 52827: File/attachment fields with special characters

XingY opened the pull request:
#### Rationale
- "multipart/form-data" is used when submitting a html form with file, instead of "application/x-www-form-urlencoded". When submitted as form, form parameters with " are converted to %22 since " is a special character for URL. This is resulting in the parameter being ignored since it cannot be matched to an existing table column.
- :: is used as delimiter for file parameters. The logic to extract parts from the delimiter is flawed in that it doesn't account for fields that contains such substring. 

#### Related Pull Requests
- <!-- list of links to related pull requests (replace this comment) -->

#### Changes
- Added MatchType.mutiPartFormData to relax column matching with encoded "
- fix logic to extract field name from file parameters

XingY committed:
Issue 52827: Cannot attach a file to a sample if the field name contains tricky characters (unclosed " or containing ::)

labkey-sweta committed:
Test updates for AutoLinkToStudyTest.testPredefinedDatasetCategories (#6559)

labkey-jeckels committed:
Issue 52917: File Browser doesn't load with non-root context path (#6602)

labkey-nicka committed:
Issue 52908: Ignore webpack:// source paths in mothership.js (#6605)

XingY committed:
Merge branch 'develop' into fb_issue52827

# Conflicts:
#	assay/src/org/labkey/assay/actions/ImportRunApiAction.java

labkey-matthewb commented on query/src/org/labkey/query/controllers/QueryController.java:
```diff
@@ -4643,26 +4643,34 @@ private void addRowAttachments(Map<String, Object> rowMap, int rowIndex, @Nullab
+                                if (!fieldCommandIndex.equals(commandIndex+""))
+                                    continue;
+                            }
+                            else
```
I think the correct fix is to remove the encoded characters here (early) rather than trying to match them later.  Ideally the encoded chars shouldn't "leak" out of code that is pulling values out of the request object.

Then we wouldn't have to handle these characters later.

XingY replied:
The challenge here is that a field might actually contain "%22". Both field" and field%22 will come in as field%22 and we don't know when to decode vs not here.

labkey-matthewb replied:
🤯

labkey-matthewb replied:
This is really an HTTP bug (it's not OK to escape " to %22 without escaping %).  So this would break any column with " in its name in a type with a file column then.  Not just the file column.

Since these columns could never map to a java property.  I'd propose changing BaseColumnInfo.getPropertyName() to replace **"** with **%_QUOTE_%** or something.  That should solve the problem by never using the " in the input name.  That would fix the dataregion case, anyway.  I'm not sure what the editable grid does.

XingY replied:
Editable grid in apps currently doesn't support file fields. The paths supported in Apps are editing detail editing and bulk ediitng. Detail editing uses `updateRows` api with formData (file, plus json data for other row data) while bulk update uses `saveRows`. For Apps, only the file field is impacted since other fields are in json. This approach of replace " with %QUOTE% could work for Apps too, but it's also more error prone. Also, it's hard to pick a substitute that absolutely won't collide with real field names, a field might just actually contain `%QUOTE%`.

labkey-matthewb commented on api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:
```diff
@@ -141,7 +141,31 @@ public static Map<String,ColumnInfo> createTableMap(TableInfo target, boolean us
+                    {
+                        if (name == null)
+                            return null;
+                        // " is encoded as %22 when content-type is "multipart/form-data"
```
```suggestion
                        // " is encoded as %22 when content-type is "multipart/form-data" (but is not otherwise encoded so decode() does not work)
```

labkey-matthewb reviewed (approved):
We can use your approach, but I would add a longer explanation about the problem with the form encoding.  We might also consider removing double-quote from generated property name.  BaseColumnInfo.getPropertyName().

labkey-jeckels committed:
Fix assorted exceptions from bots/crawlers (#6608)

* Try to track down OutOfMemoryError saving custom view

XingY committed:
Update api/src/org/labkey/api/dataiterator/DataIteratorUtil.java

Co-authored-by: labkey-matthewb <matthewb@labkey.com>

XingY committed:
bug fixes

cnathe committed:
Misc 25.5 assay result issue fixes (assay result delete rows issue 51126, assay unresolved sample lookup issue 52745) (#6609)

- AssayResultUpdateService to call syncLineage() on assay result delete rows when sample lookup column exists
- ExprColumn null check for various sample lookup fields (IsAliquot, IsPlated, StorageStatus)

XingY committed:
disallow presence of both " and %22 field names

XingY committed:
disallow presence of both " and %22 field names

XingY committed:
fix 2 errors

XingY committed:
Merge branch 'develop' into fb_issue52827

# Conflicts:
#	api/src/org/labkey/api/exp/property/DomainUtil.java

XingY committed:
fix attachment

cnathe commented on api/src/org/labkey/api/data/TableViewForm.java:
```diff
@@ -572,6 +574,26 @@ public Map<String,Object> getTypedColumns(boolean includeUntyped)
+            {
+                String quoteEncodedFieldName = getMultiPartFormFieldName(column);
+                boolean isFileColFileRemoved = false;
+                if (values.get(column.getName()) == null && File.class.equals(column.getJavaClass()) && getRequest() instanceof MultipartHttpServletRequest)
```
didn't you already check for `getRequest() instanceof MultipartHttpServletRequest` at line 577 above?

XingY replied:
Good point. Removed.

cnathe commented on api/src/org/labkey/api/data/TableViewForm.java:
```diff
@@ -572,6 +574,26 @@ public Map<String,Object> getTypedColumns(boolean includeUntyped)
+                boolean isFileColFileRemoved = false;
+                if (values.get(column.getName()) == null && File.class.equals(column.getJavaClass()) && getRequest() instanceof MultipartHttpServletRequest)
+                {
+                    MultipartHttpServletRequest request = (MultipartHttpServletRequest) getRequest();
+                    MultipartFile f = request.getFile(quoteEncodedFieldName);
+                    isFileColFileRemoved = f != null && (f.getOriginalFilename() == null || f.getOriginalFilename().isEmpty());
+                }
+
+                if (isFileColFileRemoved)
+                    values.put(column.getName(), null);
+                else
+                {
+                    Object value = getTypedValues().get(quoteEncodedFieldName);
+                    if (value != null)
+                        values.put(column.getName(), value);
+                }
```
this seems like a lot of duplicate code from what is already there below for handling the non-slash column name case (lines 599-607). it seems like the main part is the field name to use in `request.getFile()`. Could these blocks of code be collapsed and have the conditional part be using either getMultiPartFormFieldName or getFormFieldName? Then after we have the MultipartFile, the rest of the code seems pretty similar.

XingY replied:
This was how it was written in a previous commit, but that broke LKS attachment update. The different part is what value to use for file field: string or File. The only shared code is the following 2 lines so I removed the shared util in the last commit.
`
MultipartHttpServletRequest request = (MultipartHttpServletRequest) getRequest();
MultipartFile f = request.getFile(quoteEncodedFieldName);
`

cnathe commented on api/src/org/labkey/api/dataiterator/DataIteratorUtil.java:
```diff
@@ -141,7 +141,60 @@ public static Map<String,ColumnInfo> createTableMap(TableInfo target, boolean us
+                },
+        low;
+
+        public String getMatchedName(@Nullable String name)
+        {
+            return name;
+        }
```
I'm confused where this getMatchedName() is used. Couldn't that usage just use the name property directly?

cnathe replied:
or is this overridden somewhere that I am just not seeing?

XingY replied:
It's overriden in multiPartFormData

XingY committed:
code review change
